starter/src/main/java/org/teiid/spring/autoconfigure/PlatformTransactionManagerAdapter.java

```diff
@@ -160,111 +172,96 @@ public void setPlatformTransactionManager(PlatformTransactionManager platformTra
         }
     }
 
-    // When a third party JTA transaction manager is registered, this method will be
-    // called.
-    public void setJTATransactionManager(TransactionManager txnManager) {
-        this.jtaTransactionManager = txnManager;
-    }
-
     @Override
     public Transaction getTransaction() throws SystemException {
         try {
-            if (allowJTA()) {
-                return this.jtaTransactionManager.getTransaction();
-            }
-
             if (platformTransactionManager == null) {
                 return null;
             }
-            TransactionStatus status = null;
-            try {
-                status = TransactionAspectSupport.currentTransactionStatus();
-            } catch (NoTransactionException e) {
+            TransactionHolder holder = getOrCreateTransaction(false);
+            if (holder == null) {
                 return null;
             }
-            // status =
-            // platformTransactionManager.getTransaction(DEFAULT_TRANSACTION_DEFINITION);
-            synchronized (transactions) {
-                PlatformTransactionAdapter adapter = transactions.get(status);
-                if (adapter == null) {
-                    adapter = new PlatformTransactionAdapter(status);
-                    transactions.put(status, adapter);
+            return holder.transaction;
+        } catch (IllegalTransactionStateException e) {
+            return null;
+        }
+    }
+
+    public TransactionHolder getOrCreateTransaction(boolean start) {
+        TransactionStatus status = null;
+        try {
+            //Spring managed transaction
+            status = TransactionAspectSupport.currentTransactionStatus();
+        } catch (NoTransactionException e) {
+            //Teiid programatically managed transaction
+            if (start) {
+                status = platformTransactionManager.getTransaction(NEW_TRANSACTION_DEFINITION);
+                TransactionHolder holder = new TransactionHolder(status, new PlatformTransactionAdapter(status));
+                TRANSACTION_HOLDERS.set(new TransactionHolder(status, new PlatformTransactionAdapter(status)));
+                return holder;
+            } else {
+                try {
+                    status = platformTransactionManager.getTransaction(EXISTING_TRANSACTION_DEFINITION);
+                    //success means that there is one defined/associated, so we are safe to use
+                    //the thread local
+                    return TRANSACTION_HOLDERS.get();
+                } catch (TransactionException e1) {
+                    TRANSACTION_HOLDERS.remove();
                 }
-                return adapter;
             }
-        } catch (IllegalTransactionStateException e) {
+        }
+        if (status == null) {
             return null;
         }
+        synchronized (transactions) {
```

Comment:
The way our jta logic works is that it expects TransactionManager.getTransaction to return the same Transaction instance when we're under than transaction.  There are two cases that PlatformTransactionManagerAdapter now handles - JPA / Spring managed transactions, and Teiid managed.  If Spring is managing the transaction for example with `@Transactional` it will set the thread local on TransactionAspectSupport and we can track the transaction by that TransactionStatus.  If Spring is not managing the transaction, then we have to track the transaction with our own thread local.